### ISSUE TITLE
DGElasticityIntegrator

### ISSUE THREAD

**Issue by martemyev** [mfem/mfem#41]
Dear All,

There is a DGDiffusionIntegrator that complements a DiffusionIntegrator for implementation of a DG method (SIPG, NIPG, etc). I wonder how difficult it would be to create a DGElasticityIntegrator that being added to ElasticityIntegrator could be used for implementation of a DG method (same SIPG, NIPG) for problems of linear elasticity (I would use it to solve an elastic wave equation in a fractured domain, because fractures can be naturally taken into account in DG).

Thanks,
Mikhail

**Comment by tzanio** [comment]
Hi Mikhail,

I'm not sure what is the right DG formulation for elasticity, but I suspect that people have looked at that in the literature. Maybe other GitHub users or members of the team know more about this and will like to chime in?

If it is clear what the bilinear and linear forms should look like, we can give a better advice how difficult it will be to implement them. Most integrators are not a lot of code, but one has to be careful to make sure they are correct.

Cheers,
Tzanio

**Comment by martemyev** [comment]
Hi Tzanio,

I found this description of the DG for elasticity quite well explained and it's the same among all the formulations I've seen so far in literature: http://fenicsproject.org/pub/femcenter/pub/preprints/pdf/phiprint-2000-09.pdf (specifically paragraph 2. A Discontinuous Galerkin method for elasticity). It contains the formulation of the bilinear and linear forms. In the bilinear form (equation (2.6)) the first term is exactly what mfem::ElasticityIntegrator implements, so mfem::DGElasticityIntegrator would need to take care of integrals over interior and Dirichlet boundary faces.

Thanks for your help!

Cheers,
Mikhail

**Comment by tzanio** [comment]
Hi Mikhail,

Yes, with some effort, assembling (2.6) is doable with MFEM.

The interface terms without P0 are very similar to the [DGDiffusionIntegrator class](http://mfem.github.io/doxygen/html/classmfem_1_1DGDiffusionIntegrator.html), except that they are vector instead of scalar and gradient(u) should be replaced by sigma. It shouldn't be too difficult to start from that implementation and generalize it.

The P0 term is new, but can also be implemented if necessary.

Not sure if this is relevant, but you can also consider discretizing with non-conforming element. We have the low order collection [LinearNonConf3DFECollection](http://mfem.github.io/doxygen/html/classmfem_1_1LinearNonConf3DFECollection.html).

Hope this helps,
Tzanio

**Comment by martemyev** [comment]
Hi All,

I'd like to get back to this problem, and see if you could help me implement this DGElasticityIntegrator class. I started implementing it myself, but I'm stuck somewhere in the middle, and don't really know how to proceed. I'm trying to take something from DGDiffusionIntegrator and ElasticityIntegrator classes, but that confuses me even further. That's what I have so far:

`cxx
<file bilinteg.hpp>
/** Integrator for the DG form (taken from the PhD Thesis of Jonas De Basabe
    High-Order Finite Element Methods for Seismic Wave Propagation, UT Austin,
    2009, p. 23)

    - < { tau(u) }, [v] > + sigma < { tau(v) }, [u] >
    + kappa < { lambda + 2 mu } [u], [v] >,

    where tau(u) is traction, and for isotropic medium it's:
    tau_i(u) = stress_ij(u).n_j = lambda u_kk n_i + mu (u_ij + u_ji) n_j

    In other words:
    - < { stress(u).n }, [v] > + sigma < { stress(v).n }, [u] >
    + kappa < { lambda + 2 mu } [u], [v] >

    We know that
    stress(u) = lambda div(u) + 2 mu strain(u)
              = lambda div(u) + 2 mu 1/2 (grad(u) + (grad(u))^T)
              = lambda div(u) + mu (grad(u) + (grad(u))^T)

    where lambda and mu are Lame coefficient (see ElasticityIntegrator),
    u, v are the trial and test spaces, respectively. The parameters sigma and
    kappa determine the DG method to be used (when this integrator is added to
    the "broken" ElasticityIntegrator):

    * sigma = 0, IIPG (Dawson, C., Sun, S., & Wheeler, M., 2004. Compatible
    algorithms for coupled flow and transport, Computer Methods in Applied
    Mechanics and Engineering, 193(23-26), 2565-2580.)
    * sigma = -1, SIPG (Grote, M., Schneebeli, A., & Schotzau, D., 2006.
    Discontinuous Galerkin Finite Element Method for the Wave Equation, SIAM
    Journal on Numerical Analysis, 44(6), 2408-2431.)
    * sigma = 1, NIPG (Riviere, B., Wheeler, M., & Girault, V., 2001. A Priori
    Error Estimates for Finite Element Methods Based on Discontinuous
    Approximation Spaces for Elliptic Problems, SIAM Journal on Numerical
    Analysis, 39(3), 902-931.)

    This is a 'Vector' integrator, i.e. defined for FE spaces
    using multiple copies of a scalar FE space.
 */
class DGElasticityIntegrator : public BilinearFormIntegrator
{
protected:
   Coefficient *lambda, *mu;
   double sigma, kappa;

   // these are not thread-safe!
   Vector shape1, shape2, dshape1dn, dshape2dn, nor, nh, ni; 
   DenseMatrix jmat, dshape1, dshape2, mq, adjJ;

public:
   DGElasticityIntegrator(double s, double k)
      : lambda(NULL), mu(NULL), sigma(s), kappa(k) { }
   DGElasticityIntegrator(Coefficient &l, Coefficient &m, double s, double k)
      : lambda(&l), mu(&m), sigma(s), kappa(k) { }
   using BilinearFormIntegrator::AssembleFaceMatrix;
   virtual void AssembleFaceMatrix(const FiniteElement &el1,
                                   const FiniteElement &el2,
                                   FaceElementTransformations &Trans,
                                   DenseMatrix &elmat);
};
`

I tried to express the bilinear form in terms of div(u) and grad(u), so that I should be able to use existing MFEM constructions. Here we go with the bilininteg.cpp file where I borrowed the most of functionality from DGDiffusionIntegrator:

`cxx
void DGElasticityIntegrator::AssembleFaceMatrix(
   const FiniteElement &el1, const FiniteElement &el2,
   FaceElementTransformations &Trans, DenseMatrix &elmat)
{
   int dim, ndof1, ndof2, ndofs;
   bool kappa_is_nonzero = (kappa != 0.);
   double w, wq = 0., L, M;

   dim = el1.GetDim();
   ndof1 = el1.GetDof();

   nor.SetSize(dim);
   nh.SetSize(dim);
   ni.SetSize(dim);
   adjJ.SetSize(dim);

   shape1.SetSize(ndof1);
   dshape1.SetSize(ndof1, dim);
   dshape1dn.SetSize(ndof1);
   if (Trans.Elem2No >= 0)
   {
      ndof2 = el2.GetDof();
      shape2.SetSize(ndof2);
      dshape2.SetSize(ndof2, dim);
      dshape2dn.SetSize(ndof2);
   }
   else
   {
      ndof2 = 0;
   }

   ndofs = ndof1 + ndof2;
   elmat.SetSize(ndofs);
   elmat = 0.;
   if (kappa_is_nonzero)
   {
      jmat.SetSize(ndofs);
      jmat = 0.;
   }

   const IntegrationRule *ir = IntRule;
   if (ir == NULL)
   {
      // a simple choice for the integration order; is this OK?
      int order;
      if (ndof2)
      {
         order = 2*max(el1.GetOrder(), el2.GetOrder());
      }
      else
      {
         order = 2*el1.GetOrder();
      }
      ir = &IntRules.Get(Trans.FaceGeom, order);
   }

   // assemble elmat:
   // < { stress(u).n }, [v] > =
   // < { (lambda div(u) + mu (grad(u) + (grad(u))^T)).n }, [v] >
   // assemble jmat:
   // kappa < { lambda + 2 mu } [u], [v] >
`

Further code involves the integration, and here I'm stuck.

Any help is highly appreciated!

Thank you,
Mikhail

**Comment by martemyev** [comment]
To make it a bit clearer with the dimensions of vectors, `elmat` matrix is computed from this:
```
elmat: < { (lambda div(u) I + mu (grad(u) + (grad(u))^T)).n }, [v] >
```
where `I` is the identity matrix. So, I guess I'll need to compute these:
```
lambda (div(u) I).n
mu grad(u).n
mu (grad(u))^T.n
```
u is from the vector FE space for elasticity problem.
Any hint about how to implement this will be highly appreciated!

Thanks,
Mikhail

**Comment by v-dobrev** [comment]
Hi Mikhail,

Regarding your code above, I see one thing that needs fixing: the sizes of the matrices `elmat` and `jmat` must be `dim*ndofs` instead of `ndofs`.

Let's look first at what the input of `AssembleFaceMatrix(...)` is:
* the `FaceElementTransformations &Trans` structure contains the following (see also this [diagram](https://github.com/mfem/mfem/blob/master/fem/eltrans.hpp#L127)):
  - `Elem1No`, `Elem2No`: the indices of the two elements adjacent to the face we need to integrate over; if `Elem2No < 0` then there is only one adjacent element - `Elem1No`,
  - `FaceGeom`: the type of the reference face `Geometry`, e.g. `Geometry::TRIANGLE`,
  - `Elem1`, `Elem2`: these are pointers to the `ElementTransformation`s for the two adjacent elements,
  - `Face`: this is the `ElementTransformation` for the face,
  - `Loc1`, `Loc2`: these are the `IntegrationPointTransformation`s that allow you to transform an `IntegrationPoint` from the reference face to the reference spaces of `Elem1No` and `Elem2No`;
 as written at the top of the [diagram](https://github.com/mfem/mfem/blob/master/fem/eltrans.hpp#L127), we have the following relation:
```text
   Elem1(Loc1(x)) = Face(x) = Elem2(Loc2(x))
```
* the `const FiniteElement &el1`, `const FiniteElement &el2` are the `FiniteElements` corresponding to the adjacent elements - `Elem1No` and `Elem2No`,
* the `DenseMatrix &elmat` is the output computed matrix; it has a 2x2 block structure, where the blocks correspond to the two adjacent elements; in your case, block `(1,1)` has dimensions `(dim*ndof1,dim*ndof1)`, block `(1,2)` has dimensions `(dim*ndof1,dim*ndof2)`, etc.

Now to compute the integral over the face, you need to map it back to the reference face and apply quadrature - similar to [this loop](https://github.com/mfem/mfem/blob/master/fem/bilininteg.cpp#L2012):
```c++
   for (int p = 0; p < ir->GetNPoints(); p++)
   {
      const IntegrationPoint &ip = ir->IntPoint(p);
      IntegrationPoint eip1, eip2;

      Trans.Face->SetIntPoint(&ip);
      // assuming dim >= 2
      CalcOrtho(Trans.Face->Jacobian(), nor);

      Trans.Loc1.Transform(ip, eip1);
      el1.CalcShape(eip1, shape1);
      el1.CalcDShape(eip1, dshape1);
      // ...
```
At this point `nor` contains the normal vector for the current integration point, however it is not unit, the length of `nor` is exactly the weight you get when changing the variables (in the face integral) from the physical face space to the reference face space. You also have the values of all basis functions in element 1 in the `Vector` `shape1` and all their derivatives in the `DenseMatrix` `dshape1` - its size is `(ndof1,dim)`. All these values and derivatives are computed at the current integration point `ip` mapped from the face reference space to the element reference space, `eip1`. Also, the derivatives are computed with respect to the reference coordinate, so they need to be transformed with `J^{-T}` where `J` is the Jacobian matrix that can be computed using:
```c++
   Trans.Elem1->SetIntPoint(&eip1);  // set the IntegrationPoint where we want to compute J
   const DenseMatrix &J = Trans.Elem1->Jacobian();
```
This is all the data you need in order to form the `(1,1)` block for the forms
```text
< {lambda (div(u) I).n}, [v] >
< {mu grad(u).n}, [v] >
< {mu (grad(u))^T.n}, [v] >
```
I'll stop here for now, let me know if you have any questions thus far.

Veselin

**Comment by martemyev** [comment]
Hi Veselin,

Thank you for the explanation. Many things became clearer. This is my implementation of the `(1,1)` block of the initial `elmat` matrix. I moved the definition of all the objects closer to their initialization, to make it more understandable for me, so it's a bit different from the usual MFEM-way of implementation.
```cxx
void DGElasticityIntegrator::AssembleFaceMatrix(
   const FiniteElement &el1, const FiniteElement &el2,
   FaceElementTransformations &Trans, DenseMatrix &elmat)
{
   const int dim = el1.GetDim();
   const int ndof1 = el1.GetDof();

   const int ndof2 = (Trans.Elem2No >= 0) ? el2.GetDof() : 0;
   const int ndofs = ndof1 + ndof2;

   // Initially elmat corresponds to the term:
   // < { (lambda div(u) I + mu (grad(u) + (grad(u))^T)).n }, [v] >
   // But eventually, it's going to be:
   // elmat := -elmat + sigma*elmat^t + jmat
   elmat.SetSize(dim*ndofs);
   elmat = 0.;

   const bool kappa_is_nonzero = (kappa != 0.);

   // jmat corresponds to the term: kappa < { lambda + 2 mu } [u], [v] >
   DenseMatrix jmat;
   if (kappa_is_nonzero)
   {
      jmat.SetSize(dim*ndofs);
      jmat = 0.;
   }

   const IntegrationRule *ir = IntRule;
   if (ir == NULL)
   {
      // a simple choice for the integration order; is this OK?
      const int order = (ndof2) ? 2*max(el1.GetOrder(), el2.GetOrder()) : 2*el1.GetOrder();
      ir = &IntRules.Get(Trans.FaceGeom, order);
   }

   for (int p = 0; p < ir->GetNPoints(); ++p)
   {
      const IntegrationPoint &ip = ir->IntPoint(p);
      IntegrationPoint eip1;
      Trans.Loc1.Transform(ip, eip1);
      Trans.Face->SetIntPoint(&ip);

      Vector nor(dim); // normal vector (not unit)
      if (dim == 1) {
         nor(0) = 2*eip1.x - 1.0;
      } else {
         CalcOrtho(Trans.Face->Jacobian(), nor);
      }

      // values of all scalar basis functions at the integration point
      Vector shape1(ndof1);
      el1.CalcShape(eip1, shape1);

      // values derivatives of all scalar basis functions at the integration point
      DenseMatrix dshape1(ndof1, dim);
      el1.CalcDShape(eip1, dshape1);

      // weight of the quadrature rule for numerical integration
      Trans.Elem1->SetIntPoint(&eip1);
      double w1 = ip.weight / Trans.Elem1->Weight();
      if (ndof2)
         w1 *= 0.5;

      Vector ni(dim); // what is this for ?
      ni.Set(w1, nor);
      
      DenseMatrix adjJ(dim);
      CalcAdjugate(Trans.Elem1->Jacobian(), adjJ);
      
      Vector nh(dim); // unit normal vector ?
      adjJ.Mult(ni, nh);

      // Since the derivatives are computed with respect to the reference
      // coordinate, they need to be transformed with J^{-T} where J is the
      // Jacobian matrix
      DenseMatrix JinvT(dim); // J^{-T}
      CalcInverse(Trans.Elem1->Jacobian(), JinvT);
      JinvT.Transpose();

      DenseMatrix gshape1(ndof1, dim); // derivatives w.r.t. physical coordinates
      Mult(dshape1, JinvT, gshape1);

      Vector gshape1dn(ndof1); // grad(u).n
      gshape1.Mult(nh, gshape1dn);
      
      Vector divshape1(ndof1); // div(u)
      Vector one(dim);
      gshape1.Mult(one, divshape1);

      const double L1 = lambda->Eval(*Trans.Elem1, eip1);
      const double M1 = mu->Eval(*Trans.Elem1, eip1);

      for (int d = 0; d < dim; d++) {
         for (int i = 0; i < ndof1; ++i) {
            for (int j = 0; j < ndof1; ++j) {
               elmat(ndof1*d + i, ndof1*d + j) +=
                     ( (L1 * w1) * divshape1(i)*nh(d) + 
                       (M1 * w1) * gshape1dn(i) +
                       (M1 * w1) * gshape1dn(j) ) * shape1(j);
            }
         }
      }

     /* more to follow */
```

Please, let me know if I'm on a right track.

Thanks,
Mikhail

**Comment by v-dobrev** [comment]
To clarify what `ni` and `nh` are, I'll need to go into some details. Let's first look at the term
```text
< {mu grad(u).n}, [v] >
```
with scalar `u` and `v`. For the `(1,1)` block this becomes (here we use the fact that `n=normalize(nor)` is the normal pointing outside of element 1):
```text
< 0.5 mu grad(u).n, v >
```
where the `0.5` has to be replaced by `1` if there is no element 2. Using the relation
```text
grad(u)=J^{-T} hat_grad(hat_u)
```
(which is derived by differentiating `u(x)=hat_u(hat_x)`, where the `hat_` prefix denotes **reference** coordinate, function, gradient, etc) we can express
```text
grad(u).n = hat_grad(hat_u).(J^{-1} n)=(1/det(J)) hat_grad(hat_u).(adj(J) n)
```
In the last identity we used
```text
J^{-1}=(1/det(J)) adj(J).
```
We factor the scalar `1/det(J)` in order to combine it with other scalar factors which is an optimization to reduce the number of floating point operations. Taking into account the change of variables formula
```text
\int_F f.n = \int_{hat_F} f.nor
```
we can write
```text
<0.5 mu grad(u), v > = \int_{hat_F} 0.5 mu/det(J) hat_gard(hat_u).(adj(J) nor) v
```
Now, to implement this, after applying the quadrature rule, we can compute the contribution from one quadrature point using:
```c++
   mu1=mu->Eval(*Trans->Elem1, eip1);
   w1 = 0.5*ip.weight*mu1/Trans.Elem1->Weight();
   ni.Set(w1, nor); // ni = w1 nor
   adjJ.Mult(ni, nh); // nh = adj(J) ni
   dshape1.Mult(nh, dshape1dn); // dshape1dn = dshape1 nh
   for (int j = 0; j < ndof1; j++)
      for (int i = 0; i < ndof1; i++)
         elmat(i,j) += shape1(i)*dshape1dn(j);
```
This is essentially what we do in [`DGDiffusionIntegrator::AssembleFaceMatrix`](https://github.com/mfem/mfem/blob/master/fem/bilininteg.cpp#L1950).

When `u` and `v` are vectors, the term
```text
< {mu grad(u).n}, [v] >
```
gives a block-diagonal `elmat`. However, the other term
```text
< {mu grad(u)^T.n}, [v] >
```
should result in full matrix `elmat` since, e.g. for `u=(0,uy,0)` and `v=(vx,0,0)` we have
```
                           |0 * 0| |nx|
v.(grad(u)^T.n) = |v1 0 0| |0 * 0| |ny|
                           |0 * 0| |nz|
```
which is not zero.

Does this all make sense?

Veselin

**Comment by martemyev** [comment]
Thanks Veselin.

Quick question: `Trans.Elem1->Weight() = det(J)`. Is that correct?

Mikhail

**Comment by v-dobrev** [comment]
Yes, when J is square. It also works when J is 2x1, 3x1, or 3x2, i.e. when the transformation is for an edge in 2D/3D, or face in 3D. It returns the weight that you get when changing the variables from an integral over the physical element to an integral over the reference element.

**Comment by v-dobrev** [comment]
This feature has been merged and will be available in v3.3.
